garak/analyze/calibration.py

```diff
@@ -0,0 +1,120 @@
+
+        self._load_attempted = True
+
+        if os.path.isfile(calibration_filename):
+            with open(calibration_filename, "r", encoding="utf-8") as calibration_file:
+                try:
+                    self._data = json.load(calibration_file)
+                except json.JSONDecodeError as je:
+                    logging.warning(
+                        "Couldn't decode calibration JSON in %s: %s",
+                        calibration_filename,
+                        je,
+                    )
+                    return None
+                except Exception as e:  # don't stop here
+                    logging.warning("Exception during calibration data load: %s", e)
+                    return None
+        else:
+            logging.warning("Calibration path not found: %s", calibration_filename)
+            return None
+
+        if "garak_calibration_meta" in self._data:
+            self.metadata = self._data["garak_calibration_meta"]
+            del self._data["garak_calibration_meta"]
+
+        return len(self._data)
+
+    def _calc_z(self, mu: float, sigma: float, score: float) -> float:
+        zscore = (score - mu) / sigma
+        return zscore
+
+    def get_z_score(
+        self,
+        probe_module: str,
+        probe_classname: str,
+        detector_module: str,
+        detector_classname: str,
+        score: float,
+    ) -> float:
+
+        if not self._load_attempted:
+            self.load_calibration()
+
+        calibration_key = (
+            f"{probe_module}.{probe_classname}/{detector_module}.{detector_classname}"
+        )
+        zscore = None
+        if calibration_key in self._data:
+            distr = self._data[calibration_key]
+            distr["sigma"] = max(distr["sigma"], MINIMUM_STD_DEV)
+            zscore = self._calc_z(distr["mu"], distr["sigma"], score)
+        return zscore
+
+    def defcon_and_comment(self, zscore: float, defcon_comments=None):
+        if defcon_comments == None:
+            defcon_comments = ZSCORE_COMMENTS
+
+        zscore_defcon, zscore_comment = None, None
+        if zscore < ZSCORE_DEFCON_BOUNDS[0]:
+            zscore_defcon = 1
+        elif zscore < ZSCORE_DEFCON_BOUNDS[1]:
+            zscore_defcon = 2
+        elif zscore < ZSCORE_DEFCON_BOUNDS[2]:
+            zscore_defcon = 3
+        elif zscore <= ZSCORE_DEFCON_BOUNDS[3]:
+            zscore_defcon = 4
+        else:
+            zscore_defcon = 5
+        zscore_comment = defcon_comments[zscore_defcon]
+        return zscore_defcon, zscore_comment
+
+    def __init__(self, calibration_path=None) -> None:
+
+        self._data = {}
+        self.metadata = None
+
+        if calibration_path is None:
+            self.calibration_filename = (
+                _config.transient.package_dir
+                / "resources"
+                / "calibration"
+                / "calibration.json"
+            )
+        else:
+            self.calibration_filename = calibration_path
```

Comment:
Prefer early input validation of the provided value as an existing path when not `None`. A lazy load  of the data seems reasonable, however verification the value will be consumable seems appropriate.

Reply:
added validation (verification deferred to existing code invoked at load time)